RFC: Redacting parameters in back traces

TimWolla opened the pull request:
RFC: https://wiki.php.net/rfc/redact_parameters_in_back_traces

---------------------

Open Issues:
- [x] Blocked by #8070
- [x] Memory Leak: https://github.com/php/php-src/pull/7921#issuecomment-1035987956
- [x] Prevent serialization or poison the object after unserialization? https://github.com/php/php-src/pull/7921#discussion_r813743903
- [x] Type of Exception thrown from `SensitiveParameterValue::__unserialize()`: https://externals.io/message/117022

cmb69 commented:
@TimWolla, RFC karma has been granted. Best of luck with the RFC! :)

TimWolla commented:
Thanks! I've updated the initial message with:

RFC: https://wiki.php.net/rfc/redact_parameters_in_back_traces
RFC Discussion: https://externals.io/message/116853

derickr commented on ext/pdo/pdo_dbh.c:
```diff
@@ -1423,6 +1424,13 @@ void pdo_dbh_init(void)
+		2,
+		zend_ce_sensitive_parameter->name,
+		0
+	);
```
It would be nice if this could be done through ArgInfo and the stubs at some point.

TimWolla replied:
I agree, this would be nice. For now I've dropped adding the attribute to existing functions from this PR. This nicely separates the implementation of the behavior from adding the attribute to functions, making this easier to review.

TysonAndre commented on Zend/zend_builtin_functions.c:
```diff
@@ -1557,27 +1559,77 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
+						if (old_arg) {
+							ZVAL_DEREF(old_arg);
+							Z_TRY_ADDREF_P(old_arg);
+							arg = *old_arg;
```
```suggestion
							ZVAL_COPY_VALUE(&arg, old_arg);
```

nit: Doesn't affect any functionality but this macro is suggested for clearer intent and to only copy relevant fields.

TysonAndre commented on Zend/zend_attributes.c:
```diff
@@ -91,6 +94,59 @@ ZEND_METHOD(AllowDynamicProperties, __construct)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	array_init(return_value);
```
```suggestion
	RETURN_EMPTY_ARRAY();
```

Your code works, but this is slightly more efficient and returns an immutable empty array instead of allocating a brand new array.

Same for the other methods that don't populate an array later.

TysonAndre commented on Zend/zend_attributes.c:
```diff
@@ -91,6 +94,59 @@ ZEND_METHOD(AllowDynamicProperties, __construct)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	array_init(return_value);
```
```suggestion
	RETURN_EMPTY_ARRAY();
```

TysonAndre reviewed (commented):
implementation seems correct at a glance, just minor nits

TimWolla commented:
> implementation seems correct at a glance, just minor nits

@TysonAndre Thank you for the review. I've implemented the suggested changes. Unfortunately the implementation is not yet correct and I could need some handholding here:

Apparently I'm mishandling the reference counting when storing the original value within `SensitiveParameterValue`. Specifically:

```
<?php

function foo(
	#[\SensitiveParameter]
	$bar = null
)
{
	throw new \Exception('error');
}

foo(...['bar' => 'bar']);
```

Will result in:

```
Fatal error: Uncaught Exception: error in php-src/test2.php:8
Stack trace:
#0 php-src/test2.php(11): foo(Object(SensitiveParameterValue))
#1 {main}
  thrown in php-src/test2.php on line 8
[Fri Feb 11 09:42:08 2022]  Script:  'php-src/test2.php'
php-src/Zend/zend_string.h(150) :  Freeing 0x00007fad2b258b80 (32 bytes), script=php-src/test2.php
=== Total 1 memory leaks detected ===
```

I believe this is about the call to `Z_TRY_ADDREF_P`. I don't know in which cases I'll need to increase the refcount myself and in which cases this will be done for me, as I've oriented myself at the original implementation. Is it safe when I just move this call into the `else` part of `if (last_was_sensitive) {`?

TimWolla commented on Zend/tests/function_arguments/sensitive_parameter.phpt:
```diff
@@ -0,0 +1,50 @@
+    ["args"]=>
+    array(1) {
+      [0]=>
+      object(SensitiveParameterValue)#%d (0) {
```
Special casing the SensitiveParameterValue class for pretty printing does not seem useful to me. Your suggestion might or might not come naturally if there's an implementation of generics in a future RFC.

Adding the type of the inner value to the `__debugInfo()` output would be an option, but I'm not sure if that makes sense.

TimWolla commented on Zend/zend_attributes.c:
```diff
@@ -91,6 +94,59 @@ ZEND_METHOD(AllowDynamicProperties, __construct)
+		RETURN_THROWS();
+	}
+
+	zend_throw_exception(NULL, "Unserializing objects of class SensitiveParameterValue is not supported.", 0);
```
To be discussed: Which exception is appropriate.

See: https://externals.io/message/117022

TimWolla commented:
The [RFC](https://wiki.php.net/rfc/redact_parameters_in_back_traces) was [accepted](https://externals.io/message/117007#117119) and I've just cleaned up this PR for review.

I've added a list of open issues / discussion points to the initial message of this PR:

- #8070 should be reviewed and merged first.
- There's a memory leak where someone with better knowledge would need to advise me: https://github.com/php/php-src/pull/7921#issuecomment-1035987956
- The type of exception thrown in `SensitiveParameterValue::__unserialize()` is up for discussion. See also: https://externals.io/message/117022

iluuu1994 commented on Zend/zend_builtin_functions.c:
```diff
@@ -1557,27 +1559,77 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
+							i
+						);
+
+						if (attribute != NULL) {
```
`last_was_sensitive = attribute != NULL;`

iluuu1994 commented on Zend/zend_builtin_functions.c:
```diff
@@ -1587,14 +1639,44 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
+				);
+
+				if (
+					(i > call->func->op_array.num_args && last_was_sensitive)
```
This is a bit cryptic. Took me a while to understand.

* `zend_get_parameter_attribute_str` could be skipped if `i > call->func->op_array.num_args` since that will always be `NULL`
* The check should be `i >= call->func->op_array.num_args` but it's used as a side-effect to make the first superfluous argument set the `last_was_sensitive` to `0`, in combination with the fact that the variadic parameter doesn't count towards `op_array.num_args` to keep it at `1`.

```c
// The first argument after num_args can still have an attribute due to variadics, thus <=
if (i <= call->func->op_array.num_args) {
	zend_attribute *attribute = zend_get_parameter_attribute_str(
		call->func->op_array.attributes,
		"sensitiveparameter",
		sizeof("sensitiveparameter") - 1,
		i
	);
	last_was_sensitive = attribute != NULL;
}
```

Seems a bit more explicit.

Also, superfluous arguments (without variadics) are missing a test.

```php
function test(
    $non_sensitive,
    #[SensitiveParameter] $sensitive,
)
{
    debug_print_backtrace();
    var_dump(debug_backtrace());
    var_dump((new Exception)->getTrace());
}

test('foo', 'bar', 'baz');
```

TimWolla replied:
> Also, superfluous arguments (without variadics) are missing a test.

Thanks, added.

Regarding the code suggestion: I'll look into that later.

iluuu1994 commented on Zend/tests/function_arguments/sensitive_parameter.phpt:
```diff
@@ -0,0 +1,50 @@
+}
+
+test('sensitive');
+
```
Please add `?>` to all your tests, this makes them executable directly and eases debugging.

TimWolla replied:
Interesting. Didn't think of this use case, that would certainly have simplified some tests, thanks.

TimWolla commented on Zend/tests/function_arguments/sensitive_parameter_value.phpt:
```diff
@@ -0,0 +1,18 @@
+var_dump($v);
+var_dump(serialize($v));
+var_dump(json_encode($v));
+var_dump($v->getValue());
```
Good catch. I believe it should: The only way of accessing the sensitive value should be the explicit call to `->getValue()`.

Is there an existing example of how the cast to array can be controlled?

iluuu1994 replied:
I don't know this code but it looks like the VM will call the `get_properties_for` handler and passing the `ZEND_PROP_PURPOSE_ARRAY_CAST` flag, which would allow you to handle the case in whatever way is appropriate.

https://github.com/php/php-src/blob/bf32299184b69b505681ee05d63872563a2c7591/Zend/zend_vm_def.h#L6201-L6212

TimWolla replied:
Thanks, implemented in f031d1688c47b54d8934e3dd9707a4b42c92d132.

TimWolla commented on Zend/tests/function_arguments/sensitive_parameter_value_unserialize.phpt:
```diff
@@ -0,0 +1,16 @@
+var_dump(unserialize(serialize($v)));
+
+--EXPECTF--
+Fatal error: Uncaught Exception: Unserializing objects of class SensitiveParameterValue is not supported. in %ssensitive_parameter_value_unserialize.php:5
```
When writing the RFC my reasoning was that the `SensitiveParameterValue` should “just work” with existing code without needing special casing: Thus allowing serialization, but omitting the inner value.

Because this serialization is a lossy operation, running the deserialization is an undefined operation, thus it is prevented.

But now that you mention it: An exception during unserialization is not recoverable.

I see two options:

1. Disallow both serialization and unserialization.
2. Allow unserialization, but poison the unserialized object and disallow calling `->getValue()` on it.

iluuu1994 replied:
I think I'd prefer option 2 since serializing exceptions can be useful, I think Symfonys profiler makes use of that, can you confirm @nicolas-grekas? But maybe this should be mentioned on the mailing list.

nicolas-grekas replied:
Honestly I missed this part of the RFC. Being able to serialize but not unserialize looks extremely WTF to me.
I'm definitely prefer throwing on both sites.

TimWolla replied:
> But maybe this should be mentioned on the mailing list.

Mailing list thread: https://externals.io/message/117136

TysonAndre replied:
> https://www.php.net/manual/en/migration80.incompatible.php

[debug_backtrace()](https://www.php.net/manual/en/function.debug-backtrace.php) and [Exception::getTrace()](https://www.php.net/manual/en/exception.gettrace.php) will no longer provide references to arguments. It will not be possible to change function arguments through the backtrace.

That's important to keep in mind when writing code to handle SensitiveParameterValue or polyfill SensitiveParameterValue in a framework.

Anything looking for SensitiveParameterValue wouldn't find anything prior to php 8.2, though, so that isn't a concern.

I'd prefer 2 since there may be existing/new code that serializes somewhere, e.g.

- checking if exceptions are loosely equivalent
- estimating memory usage of an exception being logged
- unit testing, rare code paths for error handling that unexpectedly call serialize on something containing a trace, etc

I'd consider either better than throwing only on unserialize, I'd missed that earlier.

TimWolla replied:
As per the discussion on the ML (unanimous "disallow completely") and the comments here in the PR, I've adjusted the implementation to completely disallow serialization. As per @Danack:

> "no is temporary, yes is forever"

Thanks for voicing your opinion on this matter. I'll make sure to update the RFC with an errata section later.

TimWolla commented on Zend/zend_builtin_functions.c:
```diff
@@ -1557,27 +1559,77 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
+						if (old_arg) {
+							ZVAL_DEREF(old_arg);
+							Z_TRY_ADDREF_P(old_arg);
+							ZVAL_COPY_VALUE(&arg, old_arg);
```
Yes, this appears to pass all tests. I was hoping that it would also fix https://github.com/php/php-src/pull/7921#issuecomment-1035987956, but unfortunately it does not.

iluuu1994 replied:
I think that's due to a double `Z_TRY_ADDREF_P` here and in `ZEND_METHOD(SensitiveParameterValue, __construct)`. Try reverting my suggestion and move the `Z_TRY_ADDREF_P(&arg)` to the `else` case of `if (last_was_sensitive) {` below.

TimWolla replied:
Okay, so as I've assumed. Thanks!

TysonAndre commented on Zend/zend_attributes.c:
```diff
@@ -91,6 +100,91 @@ ZEND_METHOD(AllowDynamicProperties, __construct)
+	if (purpose == ZEND_PROP_PURPOSE_ARRAY_CAST) {
+		return NULL;
+	} else {
+		return zend_std_get_properties_for(zobj, purpose);
```
var_export/debug_zval_dump use ZEND_PROP_PURPOSE_DEBUG, var_export output may be logged

You may want to always return null, or zend_empty_array, they both behave about the same way

```c
//ext/ffi/ffi.c
static HashTable *zend_fake_get_properties(zend_object *obj) /* {{{ */
{
	return (HashTable*)&zend_empty_array;
}
``` 

```c
typedef enum _zend_prop_purpose {
	/* Used for debugging. Supersedes get_debug_info handler. */
	ZEND_PROP_PURPOSE_DEBUG,
	/* Used for (array) casts. */
	ZEND_PROP_PURPOSE_ARRAY_CAST,
	/* Used for serialization using the "O" scheme.
	 * Unserialization will use __wakeup(). */
	ZEND_PROP_PURPOSE_SERIALIZE,
	/* Used for var_export().
	 * The data will be passed to __set_state() when evaluated. */
	ZEND_PROP_PURPOSE_VAR_EXPORT,
	/* Used for json_encode(). */
	ZEND_PROP_PURPOSE_JSON,
	/* Dummy member to ensure that "default" is specified. */
	_ZEND_PROP_PURPOSE_NON_EXHAUSTIVE_ENUM
} zend_prop_purpose;
```

TimWolla replied:
Thanks, I've added var_export() and debug_zval_dump() to the test: b51d8241f4c3eecd0cb3280969b6852b0b6f792c. The latter already was protected (I assume due to `__debugInfo()`).

FWIW: `return (HashTable*)&zend_empty_array;` didn't work. That causes a segmentation fault. I've used `NULL`.

TysonAndre replied:
> FWIW: return (HashTable*)&zend_empty_array; didn't work. That causes a segmentation fault. I've used NULL.

Oh, I see why. ext/ffi/ffi.c overrides both get_properties and get_debug_info so zend_std_get_properties_for uses get_debug_info instead, which by design always returns a refcounted (non-immutable) array.

Returning null in `get_properties_for` seems to work for all use cases I checked.

(GC_PROTECT_RECURSION unconditionally modifies flags, GC_TRY_PROTECT_RECURSION)


```c
// zend_ffi_cdata_get_debug_info
		case ZEND_FFI_TYPE_FUNC:
			ht = zend_new_array(0);
			// TODO: function name ???
			*is_temp = 1;
			return ht;
```

```c
	case IS_OBJECT:
		myht = zend_get_properties_for(struc, ZEND_PROP_PURPOSE_DEBUG);
		if (myht) {
			if (GC_IS_RECURSIVE(myht)) {
				PUTS("*RECURSION*\n");
				zend_release_properties(myht);
				return;
			}
			GC_PROTECT_RECURSION(myht);
		}
```

```
==103531==                                                                                                                     
==103531== Process terminating with default action of signal 11 (SIGSEGV)                                                      
==103531==  Bad permissions for mapped region at address 0x1106CA4
==103531==    at 0x66595A: php_debug_zval_dump (var.c:353)                                                                     
==103531==    by 0x665FA9: zif_debug_zval_dump (var.c:413)                                                                     
==103531==    by 0x7C2004: ZEND_DO_ICALL_SPEC_RETVAL_UNUSED_HANDLER (zend_vm_execute.h:1250)                                   
==103531==    by 0x83FA69: execute_ex (zend_vm_execute.h:55499)                                                                
==103531==    by 0x845313: zend_execute (zend_vm_execute.h:59858)
==103531==    by 0x780073: zend_execute_scripts (zend.c:1744)  
==103531==    by 0x6C2014: php_execute_script (main.c:2535)                                                                    
==103531==    by 0x907155: do_cli (php_cli.c:965) 
==103531==    by 0x908435: main (php_cli.c:1367) 
```

TysonAndre commented on Zend/zend_attributes_arginfo.h:
```diff
@@ -71,3 +111,31 @@ static zend_class_entry *register_class_AllowDynamicProperties(void)
+
+	INIT_CLASS_ENTRY(ce, "SensitiveParameterValue", class_SensitiveParameterValue_methods);
+	class_entry = zend_register_internal_class_ex(&ce, NULL);
+	class_entry->ce_flags |= ZEND_ACC_FINAL;
```
Adding `|ZEND_ACC_FORBID_DYNAMIC_PROPERTIES` seems like it'd make sense, dynamic property support ($sensitiveParameter->dynamicProp = 123;) is going away in 9.0

TimWolla replied:
Makes sense, done.

cmb69 commented on Zend/tests/function_arguments/sensitive_parameter_correctly_captures_original.phpt:
```diff
@@ -0,0 +1,70 @@
+    \var_dump($testFrame['function']);
+    \var_dump(\count($testFrame['args']));
+    \var_dump($testFrame['args'][0]);
+    \assert($testFrame['args'][1] instanceof \SensitiveParameterValue);
```
I'd go with a `var_dump()`; alternatively, assertions could be enabled in an `--INI--` section.

TimWolla replied:
I've added theses `assert()`s primarily as IDE assistance, so that it knows that a `->getValue()` methods exists (and what it shall do). Even if assertions are disabled, the test will fail in the next line if the assertion would be violated. 

A `var_dump()` doesn't solve the original intent, but I certainly don't want to do something unusual. Would you recommend:

1. Add
```
--INI--
zend.assertions=1
assert.active=1
```
2. Remove the line completely.
3. Leave it as-is.

?

TimWolla commented:
FreeBSD failure appears unrelated, a flaky test?

> 
> =====================================================================
> FAILED TEST SUMMARY
> ---------------------------------------------------------------------
> Test function pcntl_rfork() with no wait flag. [ext/pcntl/tests/pcntl_rfork_nowait.phpt]
> =====================================================================

TimWolla commented:
Thanks to @bwoebi's assistance #8070 is resolved now and there are no known blockers.

As such this one is ready for review!

TysonAndre commented on Zend/zend_attributes.c:
```diff
@@ -315,6 +384,17 @@ void zend_register_attribute_ce(void)
+	zend_ce_sensitive_parameter_value->create_object = attributes_sensitive_parameter_value_new;
+	memcpy(&attributes_object_handlers_sensitive_parameter_value, &std_object_handlers, sizeof(zend_object_handlers));
+	attributes_object_handlers_sensitive_parameter_value.offset = XtOffsetOf(attributes_sensitive_parameter_value, std);
+	attributes_object_handlers_sensitive_parameter_value.free_obj = attributes_sensitive_parameter_value_free_obj;
```
Nit: This works correctly, but doesn't need to be overridden with the current design and may confuse people about when to use XtOffsetOf and offset.

It calls `zend_object_std_dtor` with the exact same address casted (offset=0), and zend_object_std_dtor is the default.

It'd be needed if you weren't using readonly properties but were putting things before std instead (e.g. ext/spl/spl_fixedarray.c)

cmb69 replied:
As of PHP 8.0.0, new code should better use `offset` of `XtOffsetOf`.

TysonAndre replied:
```
#ifndef XtOffsetOf
# define XtOffsetOf(s_type, field) offsetof(s_type, field)
#endif
```

I assume you mean "use `offsetof` instead of `XtOffsetOf`". Good to know that it's unconditionally defined one way

cmb69 replied:
Yeah, right, "offsetof instead of XtOffsetOf".

TimWolla replied:
Let me give the reasons behind why I did it, like I did it:

1. I've used a regular readonly property in the stub so that SensitiveParameterValue behaves like a user-land object as much as possible. It should protect against *accidental* exposure of the sensitive value, but if one _explicitly_ uses Reflection to access the internal value (for whatever reason!) then it should work as if the object was implemented in userland (e.g. in a Polyfill).

As an example, this should be possible:

```
<?php

$foo = new SensitiveParameterValue('foo');

$r = new ReflectionClass($foo);
$p = $r->getProperty('value');

var_dump($p->getValue($foo));
```

2. Protecting against the `(array)` cast is only possible for native implementations and for that I need to override `get_properties_for`. While I don't need to store any additional internal values, I've added the `attributes_sensitive_parameter_value` struct with only the `std` member for consistency with other natively implemented classes, so that the SensitiveParameterValue class is no special case. Likewise I've calculated the offsets, even if I know they are always `0`, so that adding additional struct members works properly without needing to adjust multiple locations.

If you think that I should simple use `zend_object` instead of `attributes_sensitive_parameter_value` then I can adjust the code, no problem. Just wanted to give my reasons.

Regarding `offset` vs `XtOffsetOf`: Do you have an example for that?

TimWolla replied:
@TysonAndre Okay, if there's precedent then it makes sense to simplify this. I've made the change in 2fd9a602d555473c859ce48afb95976ca0c79369.

I've also added a test for reflection in 7e94af78ad294f3e6af717cd07375c8462b97545.

If it's looking good to you I'll squash those commits into the initial commit.

TimWolla replied:
I now went ahead and rebased it, as the commit links remain valid.

TysonAndre commented on Zend/zend_attributes.c:
```diff
@@ -315,6 +384,17 @@ void zend_register_attribute_ce(void)
+	zend_ce_sensitive_parameter_value = register_class_SensitiveParameterValue();
+	zend_ce_sensitive_parameter_value->create_object = attributes_sensitive_parameter_value_new;
+	memcpy(&attributes_object_handlers_sensitive_parameter_value, &std_object_handlers, sizeof(zend_object_handlers));
+	attributes_object_handlers_sensitive_parameter_value.offset = XtOffsetOf(attributes_sensitive_parameter_value, std);
```
nit: This is always 0 and doesn't need to be overridden with the current design.

TysonAndre reviewed (approved):
lgtm, code seems correct but everything else using `zend_object std` is a struct with 2 or more properties.

Although it may make sense to leave as is and put the `zval hidden;` in the struct in a followup PR to hide it completely (and override get_gc, etc. like ext/spl/spl_fixedarray.c), so no need to change

kaznovac commented:
May I suggest putting this feature behind the ini configuration directive (defaulting to redact sensitive variables in production.ini, and show the wrapped sensitive values in development).

I find it helpful to see the config parameters (especially in docker/k8s when configuring via env/configs/secrets)

TimWolla commented:
> May I suggest putting this feature behind the ini configuration directive (defaulting to redact sensitive variables in production.ini, and show the wrapped sensitive values in development).

@kaznovac 
A php.ini knob was not part of the RFC that was voted on and accepted: https://wiki.php.net/rfc/redact_parameters_in_back_traces#phpini_defaults. As php.ini knobs are an essential part of the default RFC template, adding them now would likely require a new RFC.

That said: As the RFC author I very intentionally did not propose any php.ini knobs for this, because PHP should not change its behavior based on some global configuration. To the best of my knowledge this opinion is shared by the majority of internals participants / active voters.

iluuu1994 commented on Zend/tests/function_arguments/sensitive_parameter_correctly_captures_original.phpt:
```diff
@@ -0,0 +1,70 @@
+} catch (\Exception $e) {
+    echo $e->getMessage(), PHP_EOL;
+    $testFrame = $e->getTrace()[0];
+    \var_dump($testFrame['function']);
```
Vast majority of the tests don't prefix internal functions.

iluuu1994 commented on Zend/zend_attributes.c:
```diff
@@ -91,6 +96,55 @@ ZEND_METHOD(AllowDynamicProperties, __construct)
+	ZVAL_COPY(OBJ_PROP_NUM(Z_OBJ_P(ZEND_THIS), 0), value);
+}
+
+/* {{{ */
```
Nit: Comments added only in some places, my preference is just removing them everywhere.

iluuu1994 commented on Zend/zend_builtin_functions.c:
```diff
@@ -1557,27 +1559,68 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
+							i
+						);
+
+						last_was_sensitive = attribute != NULL;
```
Nit: This should be called `is_sensitive`, as it refers to the current and not the last parameter.

TimWolla replied:
Thanks. The variable was originally created to properly handle variadics, but has since lost its meaning during the refactoring.

While researching this I also figured out how to cleanly detect variadic parameters (`fn_flags & ZEND_ACC_VARIADIC` instead of that `<=` check` + comment) and also fixed the logic to access `.common.attributes` instead of `.op_array.attributes` which I believe should be a little cleaner as we also need to handle internal functions.

iluuu1994 commented on Zend/zend_builtin_functions.c:
```diff
@@ -1557,27 +1558,66 @@ static void debug_backtrace_get_args(zend_execute_data *call, zval *arg_array) /
-							Z_TRY_ADDREF_P(arg);
-							ZEND_HASH_FILL_SET(arg);
+						zval arg;
+						zval *old_arg = zend_hash_find_ex_ind(call->symbol_table, arg_name, 1);
```
Nit: `unredacted_arg`? `old_` makes it sound like the previous arg. Sorry I didn't mention that in the previous review.

TimWolla replied:
> Sorry I didn't mention that in the previous review.

No problem. I've opted for s/old/original/ and s/new/redacted/: https://github.com/php/php-src/compare/405914a9b8ccd27f3a2788fe61f39f68ec0c3c25..846a52d5db0541679774ebbf5de1dcf70b8a1371

TimWolla replied:
I've also just swapped "arg" and "original_arg", as that naming makes a little more sense then with the scoping.

iluuu1994 commented:
Merged as 90851977348cbb8c65fce19a1670868422414fae

Hanmac commented:
I got a Question why SensitiveValue is only ever used on Parameters and not for example Properties?

like if i got the most generic User Object

```php
class User
{
  protected $password;
}
```

and another function that takes a user object:

```php
public function doSomethingWithUser(User $user)
{
    $this->doSomethingGeneric($user);
}

// this one can't make $object as Sensitive because it might not know that it can be?
public function doSomethingGeneric(object $object)
{
   // there might be Exception
}
```


and when some Exception happen inside this other function, the user password could be exposed in the Backtrace.
One way would be to mark every function that takes a User object as Sensitive, But that might not know it if that function normally takes a more generic object.

Wouldn't it be better in such case to mark the Property for password as Sensitive so it isn't shown in the Backtrace while the User Object is?
